### PR TITLE
[WIP] TensorLayer 2.0

### PR DESCRIPTION
**Pull request by DEKHTIARJonathan**
# NETWORK API REFACTORING - TO DO LIST

## Finished work:

<details>

- **activation.py:**
  * PReluLayer: 
    - [x] refactored 
    - [x] tested
  * PRelu6Layer: 
    - [x] refactored 
    - [x] tested
  * PTRelu6Layer: 
    - [x] refactored 
    - [x] tested
- **convolution/**
  * AtrousConv1dLayer: 
    - [ ] refactored => Need to be implemented
    - [ ] tested
  * AtrousConv2dLayer: 
    - [x] refactored 
    - [x] tested
  * AtrousDeConv2dLayer: 
    - [x] refactored 
    - [x] tested
  * BinaryConv2d: 
    - [x] refactored 
    - [x] tested
  * Conv1d: 
    - [x] refactored 
    - [x] tested
  * Conv2d: 
    - [x] refactored 
    - [x] tested
  * Conv1dLayer: 
    - [x] refactored 
    - [x] tested
  * Conv2dLayer: 
    - [x] refactored 
    - [x] tested
  * Conv3dLayer: 
    - [x] refactored 
    - [x] tested
  * DeConv2d: 
    - [x] refactored 
    - [x] tested
  * DeConv3d: 
    - [x] refactored 
    - [x] tested
  * DeConv2dLayer: 
    - [x] refactored 
    - [x] tested
  * DeConv3dLayer: 
    - [x] refactored 
    - [x] tested
  * DeformableConv2d: 
    - [x] refactored 
    - [ ] tested
  * DepthwiseConv2d: 
    - [x] refactored 
    - [x] tested
  * DorefaConv2d: 
    - [x] refactored 
    - [x] tested
  * GroupConv2d: 
    - [x] refactored 
    - [x] tested
  * QuantizedConv2d: 
    - [x] refactored 
    - [x] tested
  * QuantizedConv2dWithBN: 
    - [x] refactored 
    - [x] tested
  * SeparableConv1d: 
    - [x] refactored 
    - [x] tested
  * SeparableConv2d: 
    - [x] refactored 
    - [x] tested
  * SubpixelConv1d: 
    - [x] refactored 
    - [x] tested
  * SubpixelConv2d: 
    - [x] refactored 
    - [x] tested
  * TernaryConv2d: 
    - [x] refactored 
    - [x] tested
- **dense/**
  * BinaryDenseLayer: 
    - [x] refactored 
    - [x] tested
  * DenseLayer: 
    - [x] refactored 
    - [x] tested
  * DorefaDenseLayer: 
    - [x] refactored 
    - [x] tested
  * DropconnectDenseLayer: 
    - [x] refactored 
    - [x] tested
  * QuantizedDense: 
    - [x] refactored 
    - [x] tested
  * QuantizedDenseWithBN: 
    - [x] refactored 
    - [x] tested
  * TernaryDenseLayer: 
    - [x] refactored 
    - [x] tested
- **dropout.py**
  * DropoutLayer: 
    - [x] refactored 
    - [x] tested
- **extend.py**
  * ExpandDimsLayer: 
    - [x] refactored 
    - [x] tested
  * TileLayer: 
    - [x] refactored 
    - [x] tested
- **image_resampling.py**
  * UpSampling2dLayer: 
    - [x] refactored 
    - [x] tested
  * DownSampling2dLayer: 
    - [x] refactored 
    - [x] tested
- **importer.py**
  * SlimNetsLayer:
    - [x] refactored 
    - [x] tested
  * KerasLayer: 
    - [x] refactored 
    - [x] tested
- **inputs.py**
  * InputLayer:   
    - [x] refactored 
    - [x] tested
  * OneHotInputLayer:   
    - [x] refactored 
    - [x] tested
  * Word2vecEmbeddingInputlayer:   
    - [x] refactored 
    - [x] tested
  * EmbeddingInputlayer:   
    - [x] refactored 
    - [x] tested
  * AverageEmbeddingInputlayer:  
    - [x] refactored 
    - [x] tested
- **lambda_layers.py**
  * LambdaLayer:   
    - [x] refactored  
    - [x] tested
  * ElementwiseLambdaLayer:   
    - [x] refactored  
    - [ ] tested
- **merge.py**
  * ConcatLayer:   
    - [x] refactored 
    - [x] tested
  * ElementwiseLayer:   
    - [x] refactored 
    - [ ] tested
- **noise.py**
  * GaussianNoiseLayer:   
    - [x] refactored 
    - [x] tested
- **normalization.py**
  * BatchNormLayer:   
    - [x] refactored 
    - [x] tested
  * InstanceNormLayer:   
    - [x] refactored 
    - [x] tested
  * GroupNormLayer:   
    - [ ] refactored 
    - [ ] tested
  * LayerNormLayer:   
    - [x] refactored 
    - [x] tested
  * LocalResponseNormLayer:   
    - [x] refactored 
    - [x] tested
  * SwitchNormLayer:   
    - [x] refactored 
    - [x] tested
- **padding.py**
  * PadLayer:   
    - [x] refactored 
    - [x] tested
  * ZeroPad1d:   
    - [x] refactored 
    - [x] tested
  * ZeroPad2d:   
    - [x] refactored 
    - [x] tested
  * ZeroPad3d:  
    - [x] refactored 
    - [x] tested
- **pooling/**
    * MaxPool1d:  
      - [x] refactored 
      - [x] tested
    * MaxPool2d:  
      - [x] refactored 
      - [x] tested
    * MaxPool3d:  
      - [x] refactored 
      - [x] tested
    * MeanPool1d:  
      - [x] refactored 
      - [x] tested
    * MeanPool2d:  
      - [x] refactored 
      - [x] tested
    * MeanPool3d:  
      - [x] refactored 
      - [x] tested
    * GlobalMaxPool1d: 
      - [x] refactored 
      - [x] tested
    * GlobalMaxPool2d:  
      - [x] refactored 
      - [x] tested
    * GlobalMaxPool3d:  
      - [x] refactored 
      - [x] tested
    * GlobalMeanPool1d:  
      - [x] refactored 
      - [x] tested
    * GlobalMeanPool2d: 
      - [x] refactored 
      - [x] tested
    * GlobalMeanPool3d: 
      - [x] refactored 
      - [x] tested
    * PoolLayer: 
      - [x] refactored 
      - [ ] tested
- **quantize.py**
  * SignLayer: 
    - [x] refactored 
    - [ ] tested
- **recurrent/**
  * RNNLayer: 
    - [x] refactored 
    - [x] tested
  * BiRNNLayer: 
    - [x] refactored 
    - [x] tested
  * ConvLSTMLayer: 
    - [x] refactored 
    - [x] tested
  * DynamicRNNLayer: 
    - [x] refactored 
    - [x] tested
  * BiDynamicRNNLayer: 
    - [x] refactored 
    - [x] tested
  * Seq2Seq:   
    - [x] refactored 
    - [ ] tested  # => Not Tested, need to be in TL Repo  https://github.com/tensorlayer/seq2seq-chatbot/pull/17
- **reshape.py**
  * FlattenLayer: 
    - [x] refactored 
    - [x] tested
  * ReshapeLayer: 
    - [x] refactored 
    - [x] tested
  * TransposeLayer: 
    - [x] refactored 
    - [x] tested
- **scale.py**
  * ScaleLayer: 
    - [x] refactored 
    - [x] tested
- **spatial_transformer.py**
  * SpatialTransformer2dAffineLayer: 
    - [x] refactored 
    - [ ] tested
- **stack.py**
  * StackLayer: 
    - [x] refactored 
    - [ ] tested
  * UnStackLayer: => Need to be checked! Not working
    - [ ] refactored 
    - [ ] tested
- **time_distribution.py**
  * TimeDistributedLayer: 
    - [x] refactored 
    - [ ] tested
</details>

## Unittests Status:

- [x] test_activations.py
- [x] test_array_ops.py
- [x] test_decorators.py
- [ ] test_documentation.py
- [ ] test_layers_activation.py
- [ ] test_layers_basic.py
- [ ] test_layers_convolution_1d.py
- [ ] test_layers_convolution_2d.py
- [ ] test_layers_convolution_3d.py
- [ ] test_layers_convolution_deformable.py
- [ ] test_layers_core.py
- [ ] test_layers_extend.py
- [ ] test_layers_flow_control.py
- [ ] test_layers_importer.py
- [ ] test_layers_merge.py
- [ ] test_layers_normalization.py
- [ ] test_layers_padding.py
- [ ] test_layers_pooling.py
- [ ] test_layers_recurrent.py
- [ ] test_layers_reshape.py
- [ ] test_layers_spatial_transformer.py
- [ ] test_layers_stack.py
- [ ] test_layers_super_resolution.py
- [ ] test_layers_time_distributed.py
- [x] test_logging.py
- [x] test_logging_hyperdash.py
- [ ] test_mnist_simple.py
- [ ] test_models.py
- [x] test_network_custom_2d.py
- [x] test_network_custom_input_layers.py
- [x] test_network_custom_multiple_inputs.py
- [x] test_network_custom_multiple_outputs.py
- [x] test_network_sequential_1d.py
- [x] test_network_sequential_2d.py
- [x] test_network_sequential_3d.py
- [x] test_network_sequential_rnn.py
- [ ] test_optimizer_amsgrad.py
- [ ] test_pydocstyle.py
- [ ] test_reuse_mlp.py
- [ ] test_tf_layers.py
- [x] test_timeout.py
- [ ] test_utils_predict.py
- [x] test_yapf_format.py

### Layers

- **contrib**
  * ROIPoolingLayer:   
    - [x] refactored 
    - [ ] tested
- **convolution/**
  * AtrousConv1dLayer:
    - [ ] refactored => Need to be implemented
    - [ ] tested
  * DeformableConv2d:
    - [x] refactored
    - [ ] tested
- **lambda_layers.py**
  * ElementwiseLambdaLayer:
    - [x] refactored
    - [ ] tested
- **merge.py**
  * ElementwiseLayer:
    - [x] refactored
    - [ ] tested
- **normalization.py**
  * GroupNormLayer:
    - [ ] refactored
    - [ ] tested
- **pooling/**
    * PoolLayer:
      - [x] refactored
      - [ ] tested
- **quantize.py**
  * SignLayer:
    - [x] refactored
    - [ ] tested
- **spatial_transformer.py**
  * SpatialTransformer2dAffineLayer:
    - [x] refactored
    - [ ] tested
- **recurrent/**
  * Seq2Seq:
    - [x] refactored
    - [ ] tested  # => Not Tested, need to be in TL Repo  https://github.com/tensorlayer/seq2seq-chatbot/pull/17
- **stack.py**
  * StackLayer:
    - [x] refactored
    - [ ] tested
  * UnStackLayer: => Need to be checked! Not working
    - [ ] refactored
    - [ ] tested
- **time_distribution.py**
  * TimeDistributedLayer:
    - [x] refactored
    - [ ] tested

### TensorLayer Hub

- **tl.models** => became **tl.hub**
  - [ ] VGG16
  - [ ] VGG19
  - [ ] MobileNet
  - [ ] SqueezeNet

### Examples
- **basic_tutorials**
    - [ ] refactored 
    - [ ] tested
- **data_process**
    - [ ] refactored 
    - [ ] tested
- **database**
    - [ ] refactored 
    - [ ] tested
- **deprecated_tutorials**
    - [ ] refactored 
    - [ ] tested
- **distributed_training**
    - [ ] refactored 
    - [ ] tested
- **keras_tfslim**
    - [ ] refactored 
    - [ ] tested
- **pretrained_cnn**
    - [ ] refactored 
    - [ ] tested
- **quantized_net**
    - [ ] refactored 
    - [ ] tested
- **reinforcement_learning**
    - [ ] refactored 
    - [ ] tested
- **text_classification**
    - [ ] refactored 
    - [ ] tested
- **text_generation**
    - [ ] refactored 
    - [ ] tested
- **text_ptb**
    - [ ] refactored 
    - [ ] tested
- **text_word_embedding**
    - [ ] refactored 
    - [ ] tested

## More TODO
- [ ] add `scope_name` into the layer name
- [ ] don't check duplicated layer name in core layer

Test code:

```python
plh = tf.placeholder(tf.float16, (100, 32))
net = tl.layers.InputLayer(name='in')(plh)
with tf.variable_scope('test'):
    net = tl.layers.DenseLayer(n_units=50, act=tf.nn.relu, name="dense")(net)
with tf.variable_scope('test', reuse=True):
    net = tl.layers.DenseLayer(n_units=50, act=tf.nn.relu, name="dense")(net)
print(net['test/dense'])
print(net['test/dense_2'])
assert len(net.all_weights) == 2
print(net.all_params)  # give a warning, but still works
```

How to do with manual-compile?:

```python
with tf.variable_scope('test'):
    model.add(tl.layers.DenseLayer(n_units=50, act=tf.nn.relu, name="seq_layer_9"))
with tf.variable_scope('test', reuse=True):
    model.add(tl.layers.DenseLayer(n_units=50, act=tf.nn.relu, name="seq_layer_9"))
plh = tf.placeholder(tf.float16, (100, 32))
net = model.compile(plc)
print(net['test/seq_layer_9'])
```

- [ ] `_params` and `_variables` are everywhere ....
   - `assign_params`, `get_variables_with_name`, `print_params`

### CODE REVIEW

**Comment by 2wins** [comment]
**Related Issue**: `NCHW/NHWC`, `channels_first/last` (Issue #561, PR #640)
We need to refer to comments on the posts.
